src/detours.cpp

```diff
@@ -334,14 +362,72 @@ struct _DETOUR_TRAMPOLINE
     PBYTE           pbRemain;       // first instruction after moved code. [free list]
     PBYTE           pbDetour;       // first instruction of detour function.
     BYTE            rbCodeIn[8];    // jmp [pbDetour]
+#ifdef DETOURS_SUPPORT_CONTEXT_CALLBACK
+    BYTE            rbContext[42];  // context callback code (calls context function w/ void* param)
+    PBYTE           pbContext;      // first instruction of context callback code (rbSetContext) if used
+    PF_DETOUR_INVOKE_CONTEXT_CALLBACK pfContextCallback;    // context callback used as indirect jmp in rbContext
+#endif // DETOURS_SUPPORT_CONTEXT_CALLBACK
 };
 
+#ifdef DETOURS_SUPPORT_CONTEXT_CALLBACK
+C_ASSERT(sizeof(_DETOUR_TRAMPOLINE) == 160);
+#else
 C_ASSERT(sizeof(_DETOUR_TRAMPOLINE) == 96);
+#endif // DETOURS_SUPPORT_CONTEXT_CALLBACK
 
 enum {
     SIZE_OF_JMP = 5
 };
 
+#ifdef DETOURS_SUPPORT_CONTEXT_CALLBACK
+
+inline PBYTE detour_gen_context_callback(PBYTE pbCode, PF_DETOUR_INVOKE_CONTEXT_CALLBACK* ppfContextCallback, PVOID pContextParam)
+{
+    PBYTE pbJmpSrc = pbCode + 26;
+
+    // Backup volatile registers that are used to pass __fastcall params.
+    // If the set context function performed floating point operations, 
+    // registers XMM0, XMM1, XMM2, and XMM3 would also have to be backed 
+    // up and restored. In the attempt to keep this code generation as simple 
+    // as possible, floating point operations in the set context function isn't 
+    // currently supported.
+    *pbCode++ = 0x51;   // push rcx (backup rcx)
+    *pbCode++ = 0x52;   // push rdx (backup rdx)
+    *pbCode++ = 0x41;   // push r8 (backup r8)
+    *pbCode++ = 0x50;
+    *pbCode++ = 0x41;   // push r9 (backup r9)
+    *pbCode++ = 0x51;
+
+    *pbCode++ = 0x48;   // sub rsp,20h (required by x64 __fastcall)
+    *pbCode++ = 0x83;
+    *pbCode++ = 0xEC;
+    *pbCode++ = 0x20;
+
+    *pbCode++ = 0x48;   // mov rcx,imm64 (param)
+    *pbCode++ = 0xB9;
+    *((INT64*&)pbCode)++ = (INT64)(pContextParam);
+
+    *pbCode++ = 0xFF;   // call [+imm32] (pfContextCallback)
+    *pbCode++ = 0x15;
+    *((INT32*&)pbCode)++ = (INT32)((PBYTE)ppfContextCallback - pbJmpSrc);
+
+    *pbCode++ = 0x48;   // add rsp,20h (required by x64 __fastcall)
+    *pbCode++ = 0x83;
+    *pbCode++ = 0xC4;
+    *pbCode++ = 0x20;
+
+    *pbCode++ = 0x41;   // pop r9 (restore r9)
+    *pbCode++ = 0x59;
+    *pbCode++ = 0x41;   // pop r8 (restore r8)
+    *pbCode++ = 0x58;
+    *pbCode++ = 0x5A;   // pop rdx (restore rdx)
+    *pbCode++ = 0x59;   // pop rcx (restore rcx)
```

Comment:
Likely an ABI violation.
Detours already violates ABI.
This likely makes it worse.
I understand it is under ifdef and that makes it far more acceptable, but I do wonder how it impacts your scenario too.

Does debugging successfully show stack for you?
Have you any scenarios that raise an exception with this code on stack?

Would it be better to have kinda templatized C detours to fetch the context from thread locals?

Reply:
For our use, the context function is always this:
```C++
void __stdcall set_hook_context(void* context)
{
	if (s_hook_context_tls_index_ != TLS_OUT_OF_INDEXES)
		TlsSetValue(s_hook_context_tls_index_, context);
}
```
`TlsSetValue` could theoretically raise an exception, although I don't think the current implementation would for an allocated index.

I don't know about debugging; this is a legacy component for us and it would take some time to figure that out.

As I understand it (which is admittedly not thoroughly), the goal is to effectively add another parameter to the detoured function.  The value of this parameter is determined at attach time, not compile time.  I have not been able to think of another way to handle the scenario that doesn't involve emitting code at runtime.  I suppose with enough effort (and possible assistance from the attach caller), one could actually implement this as adding another parameter rather than stashing it in TLS.  I'm not sure how difficult/feasible that would be, nor can I predict the potential impact.

Is the primary concern the lack of function table entries?  Is there enough referential data to implement the callback for `RtlInstallFunctionTableCallback`, or would that require significant investment?

Reply:
Ideally you could get stackwalks in a debugger too, but you appear to break that. I guess nobody really cares.

Reply:
You are aware of TlsGetValue / TlsSetValue's tendency to clear last error, I assume?

Reply:
You should try to use declspec(thread).